core-processor/src/main/java/io/micronaut/expressions/parser/CompoundEvaluatedEvaluatedExpressionParser.java

```diff
@@ -0,0 +1,145 @@
+ */
+package io.micronaut.expressions.parser;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.expressions.parser.ast.ExpressionNode;
+import io.micronaut.expressions.parser.ast.literal.StringLiteral;
+import io.micronaut.expressions.parser.ast.types.TypeIdentifier;
+import io.micronaut.expressions.parser.ast.collection.NewArray;
+import io.micronaut.expressions.parser.ast.operator.binary.math.AddOperator;
+import io.micronaut.expressions.parser.exception.ExpressionParsingException;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static io.micronaut.core.expression.EvaluatedExpression.EXPRESSION_PREFIX;
+
+/**
+ * This parser is used to split complex expression into multiple
+ * single expressions if necessary and delegate each atomic expression
+ * parsing to separate instance of {@link SingleEvaluatedEvaluatedExpressionParser},
+ * then combining simple expressions parsing results.
+ *
+ * @author Sergey Gavrilov
+ * @since 4.0.0
+ */
+@Internal
+public final class CompoundEvaluatedEvaluatedExpressionParser implements EvaluatedExpressionParser {
+
+    private final Object expression;
+
+    public CompoundEvaluatedEvaluatedExpressionParser(Object expression) {
+        this.expression = expression;
+    }
+
+    @Override
+    public ExpressionNode parse() throws ExpressionParsingException {
+        return parseCompoundExpression(expression);
+    }
+
+    private ExpressionNode parseCompoundExpression(Object expression) {
+        // if expression doesn't have prefix, the whole string is treated as expression
+        if (expression instanceof String str && !str.contains(EXPRESSION_PREFIX)) {
+            return new SingleEvaluatedEvaluatedExpressionParser(str).parse();
+        }
+
+        return parseTemplateExpression(expression);
+    }
+
+    private ExpressionNode parseTemplateExpression(Object expression) {
+        if (expression instanceof String str) {
+            List<ExpressionNode> expressionParts =
+                splitExpressionParts(str).stream()
+                    .map(this::prepareExpressionPart)
+                    .map(SingleEvaluatedEvaluatedExpressionParser::new)
+                    .map(SingleEvaluatedEvaluatedExpressionParser::parse)
+                    .toList();
+
+            if (expressionParts.size() == 1) {
+                return expressionParts.get(0);
+            } else {
+                return expressionParts.stream()
+                           .reduce(new StringLiteral(""), AddOperator::new);
+            }
+        } else if (expression instanceof String[] strArray) {
+            List<ExpressionNode> arrayNodes =
+                Arrays.stream(strArray)
+                    .map(this::parseTemplateExpression)
+                    .toList();
+
+            return buildArrayOfExpressions(arrayNodes);
+        } else {
+            throw new ExpressionParsingException("Can not parse expression: " + expression);
```

Comment:
perhaps move this validation to the constructor if only strings and string arrays are supported. In fact maybe add 2 constructors one that takes a string and another that takes a string array